zerver/lib/streams.py

```diff
@@ -175,15 +176,15 @@ def create_streams_if_needed(
 def check_stream_name(stream_name: str) -> None:
     if stream_name.strip() == "":
         raise JsonableError(_("Invalid stream name '{}'").format(stream_name))
+
     if len(stream_name) > Stream.MAX_NAME_LENGTH:
         raise JsonableError(
             _("Stream name too long (limit: {} characters).").format(Stream.MAX_NAME_LENGTH)
         )
     for i in stream_name:
-        if ord(i) == 0:
-            raise JsonableError(
-                _("Stream name '{}' contains NULL (0x00) characters.").format(stream_name)
-            )
+        unicode_category = unicodedata.category(i)
+        if unicode_category in ["Cc", "Cs", "Cn"]:
```

Comment:
Hmm, shouldn't we extract a helper function for validating these characters, to avoid duplicating the same logic with b3c58f454f0dce8a88c696305945bac41f9786bc?

Reply:
@timabbott @mateuszmandera This seems like a good idea, should we add this in `zerver/lib/validator.py`?

Reply:
I'd put it in a new `zerver/lib/string_validation.py`; `validator.py` is a library with a very specific purpose.

Reply:
When you're doing this refactoring, zerver/migrations/0371_invalid_characters_in_topics.py has an additional use of this pattern now as well.

Reply:
Yes, though we should just add a `# Copy of {function_name}` comment there rather than having it call into the new function; since if we change the set of character classes, we need to repeat the migration, rather than changing one that has already been run on some systems.

Reply:
@timabbott @alexmv If we create a common function for string validation, how can we generalize the error messages, Like we should pass it with the arguments of function created in `zerver/lib/string_validation.py`
```python
def check_invalid_string(var: str, error: str) -> None:
```
or else we just pass the object name such as `stream name`, `topic name` or `user group name`. 
```python
raise JsonableError(_("Invalid unicode characters in {}").format(object_name))
```

Reply:
@madrix01 Another alternative could be to raise e.g. `ValidationError` in the character validation function, and the caller can catch it and re-raise its own JsonableError.

Reply:
@mateuszmandera then we would have to add conditionals for different types of errors, in different files, is it right? Can you give an example?